File: mcp_server/src/sts2_mcp/state_views.py
"""Two read-only projections over a `/state` payload: a run summary and a state diff.

Both answer a question an agent asks constantly and had to re-derive itself:

- **`run_summary`** is "where am I in this run" in one call, instead of walking `run.deck`,
  `run.relics`, `run.potions`, and the party block on every decision.
- **`diff_state`** is "what changed" between two snapshots, which is the first thing anyone needs
  when an action's result is not what they expected.

Both are pure functions over the raw `/state` shape, so they are exercised without a game and the
C# native surface can mirror them field for field -- the raw payload's names are documented in
`docs/api.md` and do not move, while the compact `agent_view` renames many of them (see the rename
table there), which is exactly the kind of drift a mirror should not have to track.

Neither function invents a value. A field the payload does not carry is reported as `None`, and a
run that is not present answers `None` rather than an empty summary, so a caller can tell "not in a
run" from "in a run with nothing in it".
"""

from __future__ import annotations

from typing import Any

# A diff of two whole `/state` payloads can be enormous (a combat payload with a full hand, every
# enemy intent, the entire map graph). The cap keeps one call from flooding a model's context; the
# caller reads `truncated` and narrows the comparison instead of silently receiving a partial diff.
MAX_DIFF_ENTRIES = 200

# Depth is capped for the same reason: the map's `nodes[].children[]` nests unbounded in principle,
# and an unbounded walk would turn a debugging tool into a denial of service.
MAX_DIFF_DEPTH = 12


def _as_mapping(value: Any) -> dict[str, Any] | None:
    return value if isinstance(value, dict) else None


def _count(value: Any) -> int:
    return len(value) if isinstance(value, list) else 0


def _occupied_potions(potions: Any) -> int:
    if not isinstance(potions, list):
        return 0
    return sum(1 for potion in potions if isinstance(potion, dict) and potion.get("occupied"))


def run_summary(state: Any) -> dict[str, Any] | None:
    """A one-call summary of the current run, or None when the payload carries no run.

    Field names are the raw `/state` ones, so they can be looked up in `docs/api.md` rather than
    guessed. Counts are counts of what the payload actually holds: a deck of 0 is a real answer for
    a run whose deck payload is empty, and an absent `run` is not.
    """
    state_mapping = _as_mapping(state)
    run = _as_mapping(state_mapping.get("run")) if state_mapping is not None else None
    if run is None:
        return None

    party = run.get("players") if isinstance(run.get("players"), list) else []
    return {
        "character_id": run.get("character_id"),
        "character_name": run.get("character_name"),
        "floor": run.get("floor"),
        "act_id": run.get("act_id"),
        "boss_id": run.get("boss_id"),
        "ascension": run.get("ascension"),
        "current_hp": run.get("current_hp"),
        "max_hp": run.get("max_hp"),
        "gold": run.get("gold"),
        "max_energy": run.get("max_energy"),
        "deck_size": _count(run.get("deck")),
        "relic_count": _count(run.get("relics")),
        "potion_count": _count(run.get("potions")),
        "potions_occupied": _occupied_potions(run.get("potions")),
        "party": [
            {
                "player_id": member.get("player_id"),
                "is_local": member.get("is_local"),
                "is_alive": member.get("is_alive"),
                "current_hp": member.get("current_hp"),
                "max_hp": member.get("max_hp"),
                "gold": member.get("gold"),
            }
            for member in party
            if isinstance(member, dict)
        ],
    }


def _leaf(value: Any) -> tuple[str, Any]:
    """A scalar as (kind, value).

    The kind is part of the comparison on purpose. JSON `"12"` and `12` are different facts about a
    payload, and in Python `True == 1`, so comparing bare values would report "no change" for a
    boolean that turned into a number. The C# mirror tags leaves the same way.
    """
    if value is None:
        return ("null", None)
    if isinstance(value, bool):
        return ("bool", value)
    if isinstance(value, (int, float)):
        return ("number", value)
    return ("string", str(value))


def _flatten(value: Any, path: str, depth: int, out: dict[str, tuple[str, Any]]) -> None:
    if depth > MAX_DIFF_DEPTH:
        out[path] = ("string", "<max depth>")
        return

    if isinstance(value, dict):
        if not value:
            out[path] = ("string", "{}")
            return
        for key in value:
            child = f"{path}.{key}" if path else str(key)
            _flatten(value[key], child, depth + 1, out)
        return

    if isinstance(value, list):
        # Lists are compared by length and by element, so an index that appeared or vanished is a
        # change of its own rather than a reshuffle of every later index.
        out[f"{path}[]"] = ("string", f"len={len(value)}")
        for index, item in enumerate(value):
            _flatten(item, f"{path}[{index}]", depth + 1, out)
        return

    out[path] = _leaf(value)


def _leaves(value: Any) -> dict[str, tuple[str, Any]]:
    flat: dict[str, tuple[str, Any]] = {}
    _flatten(value, "", 0, flat)
    return flat


def diff_state(before: Any, after: Any, *, limit: int = MAX_DIFF_ENTRIES) -> dict[str, Any]:
    """The paths that differ between two `/state` payloads, oldest value first.

    Each entry names the path, the value before, and the value after; a path on one side only is
    reported with the missing side as None. `truncated` says the cap was reached, so an empty
    `changes` list means "no difference" and a truncated one never reads as one.
    """
    cap = max(1, int(limit))
    before_leaves = _leaves(before)
    after_leaves = _leaves(after)

    changes: list[dict[str, Any]] = []
    truncated = False
    paths = sorted(set(before_leaves) | set(after_leaves))
    for path in paths:
        has_before = path in before_leaves
        has_after = path in after_leaves
        if has_before and has_after and before_leaves[path] == after_leaves[path]:
            continue

        changes.append(
            {
                "path": path,
                "before": before_leaves[path][1] if has_before else None,
                "after": after_leaves[path][1] if has_after else None,
            }
        )

        if len(changes) >= cap:
            truncated = len(changes) < len(paths)
            break

    return {
        "changes": changes,
        "change_count": len(changes),
        "truncated": truncated,
        "limit": cap,
    }

File: STS2AIAgent/Agent/StateViews.cs
using System.Text.Json;

namespace STS2AIAgent.Agent;

/// <summary>
/// Two read-only projections over a raw <c>/state</c> payload: a run summary and a state diff.
/// </summary>
/// <remarks>
/// This mirrors <c>mcp_server/src/sts2_mcp/state_views.py</c> field for field, the way ADR 0002
/// requires of the two MCP surfaces. Both work on the <b>raw</b> payload rather than the compact
/// <c>agent_view</c>, because the raw field names are the documented ones in <c>docs/api.md</c> and
/// do not move, while the compact view renames many of them (it has its own rename table there) --
/// a mirror that read the compact shape would have to track that table too.
///
/// Neither projection invents a value. A field the payload does not carry serializes as null, and a
/// payload with no <c>run</c> answers with a null summary, so a caller can tell "not in a run" from
/// "in a run with nothing in it".
/// </remarks>
internal static class StateViews
{
    /// <summary>
    /// A diff of two whole payloads can be enormous (a full hand, every intent, the map graph), so
    /// the cap keeps one call from flooding a model's context; the caller reads <c>truncated</c> and
    /// narrows the comparison instead of silently receiving a partial diff.
    /// </summary>
    public const int MaxDiffEntries = 200;

    /// <summary>The map graph nests in principle without bound; a debugging tool must still terminate.</summary>
    public const int MaxDiffDepth = 12;

    /// <summary>
    /// A one-call summary of the current run, or null when the payload carries no run.
    /// </summary>
    public static object? BuildRunSummary(JsonElement state)
    {
        if (state.ValueKind != JsonValueKind.Object ||
            !state.TryGetProperty("run", out var run) ||
            run.ValueKind != JsonValueKind.Object)
        {
            return null;
        }

        var potions = ArrayOf(run, "potions");
        var occupied = 0;
        foreach (var potion in potions)
        {
            if (potion.ValueKind == JsonValueKind.Object &&
                potion.TryGetProperty("occupied", out var isOccupied) &&
                isOccupied.ValueKind == JsonValueKind.True)
            {
                occupied++;
            }
        }

        var party = new List<object>();
        foreach (var member in ArrayOf(run, "players"))
        {
            if (member.ValueKind != JsonValueKind.Object)
            {
                continue;
            }

            party.Add(new
            {
                player_id = Text(member, "player_id"),
                is_local = Bool(member, "is_local"),
                is_alive = Bool(member, "is_alive"),
                current_hp = Number(member, "current_hp"),
                max_hp = Number(member, "max_hp"),
                gold = Number(member, "gold")
            });
        }

        return new
        {
            character_id = Text(run, "character_id"),
            character_name = Text(run, "character_name"),
            floor = Number(run, "floor"),
            act_id = Text(run, "act_id"),
            boss_id = Text(run, "boss_id"),
            ascension = Number(run, "ascension"),
            current_hp = Number(run, "current_hp"),
            max_hp = Number(run, "max_hp"),
            gold = Number(run, "gold"),
            max_energy = Number(run, "max_energy"),
            deck_size = ArrayOf(run, "deck").Count,
            relic_count = ArrayOf(run, "relics").Count,
            potion_count = potions.Count,
            potions_occupied = occupied,
            party
        };
    }

    /// <summary>
    /// The paths that differ between two payloads; each entry carries the value before and after,
    /// with null for a side that does not have the path at all.
    /// </summary>
    public static object BuildStateDiff(JsonElement before, JsonElement after, int limit = MaxDiffEntries)
    {
        var cap = Math.Max(1, limit);
        var beforeLeaves = Flatten(before);
        var afterLeaves = Flatten(after);

        var paths = new SortedSet<string>(StringComparer.Ordinal);
        paths.UnionWith(beforeLeaves.Keys);
        paths.UnionWith(afterLeaves.Keys);

        var changes = new List<object>();
        var truncated = false;
        foreach (var path in paths)
        {
            beforeLeaves.TryGetValue(path, out var beforeValue);
            afterLeaves.TryGetValue(path, out var afterValue);
            var hasBefore = beforeLeaves.ContainsKey(path);
            var hasAfter = afterLeaves.ContainsKey(path);

            if (hasBefore && hasAfter && beforeValue == afterValue)
            {
                continue;
            }

            changes.Add(new
            {
                path,
                before = hasBefore ? beforeValue.Value : null,
                after = hasAfter ? afterValue.Value : null
            });

            if (changes.Count >= cap)
            {
                truncated = changes.Count < paths.Count;
                break;
            }
        }

        return new
        {
            changes,
            change_count = changes.Count,
            truncated,
            limit = cap
        };
    }

    /// <summary>
    /// A scalar as (kind, value). The kind is part of the comparison on purpose: JSON <c>"12"</c>
    /// and <c>12</c> are different facts about a payload, and the Python mirror cannot rely on bare
    /// equality either (<c>True == 1</c> there), so both tag leaves the same way.
    /// </summary>
    private readonly record struct Leaf(string Kind, string Text, object? Value);

    /// <summary>Every scalar path in the payload, tagged so the comparison is type-aware.</summary>
    private static Dictionary<string, Leaf> Flatten(JsonElement value)
    {
        var flat = new Dictionary<string, Leaf>(StringComparer.Ordinal);
        FlattenInto(value, string.Empty, 0, flat);
        return flat;
    }

    private static void FlattenInto(JsonElement value, string path, int depth, Dictionary<string, Leaf> flat)
    {
        if (depth > MaxDiffDepth)
        {
            flat[path] = new Leaf("string", "<max depth>", "<max depth>");
            return;
        }

        switch (value.ValueKind)
        {
            case JsonValueKind.Object:
                var wroteChild = false;
                foreach (var property in value.EnumerateObject())
                {
                    wroteChild = true;
                    var child = path.Length == 0 ? property.Name : path + "." + property.Name;
                    FlattenInto(property.Value, child, depth + 1, flat);
                }

                if (!wroteChild)
                {
                    flat[path] = new Leaf("string", "{}", "{}");
                }

                return;

            case JsonValueKind.Array:
                // Lists are compared by length and by element, so an index that appeared or vanished
                // is a change of its own rather than a reshuffle of every later index.
                var length = "len=" + value.GetArrayLength();
                flat[path + "[]"] = new Leaf("string", length, length);
                var index = 0;
                foreach (var item in value.EnumerateArray())
                {
                    FlattenInto(item, path + "[" + index + "]", depth + 1, flat);
                    index++;
                }

                return;

            default:
                flat[path] = Scalar(value);
                return;
        }
    }

    private static Leaf Scalar(JsonElement value)
    {
        switch (value.ValueKind)
        {
            case JsonValueKind.String:
                var text = value.GetString() ?? string.Empty;
                return new Leaf("string", "s:" + text, text);
            case JsonValueKind.Number:
                return new Leaf("number", "n:" + value.GetRawText(), value.GetDouble());
            case JsonValueKind.True:
                return new Leaf("bool", "b:true", true);
            case JsonValueKind.False:
                return new Leaf("bool", "b:false", false);
            default:
                return new Leaf("null", "z:", null);
        }
    }

    private static IReadOnlyList<JsonElement> ArrayOf(JsonElement parent, string name)
    {
        if (!parent.TryGetProperty(name, out var value) || value.ValueKind != JsonValueKind.Array)
        {
            return Array.Empty<JsonElement>();
        }

        var items = new List<JsonElement>();
        foreach (var item in value.EnumerateArray())
        {
            items.Add(item);
        }

        return items;
    }

    private static string? Text(JsonElement parent, string name)
    {
        return parent.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.String
            ? value.GetString()
            : null;
    }

    private static bool? Bool(JsonElement parent, string name)
    {
        if (!parent.TryGetProperty(name, out var value))
        {
            return null;
        }

        return value.ValueKind switch
        {
            JsonValueKind.True => true,
            JsonValueKind.False => false,
            _ => null
        };
    }

    private static double? Number(JsonElement parent, string name)
    {
        return parent.TryGetProperty(name, out var value) &&
               value.ValueKind == JsonValueKind.Number &&
               value.TryGetDouble(out var number)
            ? number
            : null;
    }
}
